PolicyExecutorTest.testInvokeAnyTimedInterruptWaitForEnqueue: AssertionError: Should have been interrupted

The following failure occurs intermittently in automated test,

junit.framework.AssertionFailedError: 2017-10-08-12:48:01:416 The response did not contain "[SUCCESS]".  Full output is:"
ERROR: Caught exception attempting to call test method testInvokeAnyTimedInterruptWaitForEnqueue on servlet web.PolicyExecutorServlet
java.lang.AssertionError: Should have been interrupted. Instead: true
	at web.PolicyExecutorServlet.testInvokeAnyTimedInterruptWaitForEnqueue(PolicyExecutorServlet.java:3306)

The test case constrains maxConcurrency to 1, and maxQueueSize to 1, with maxWaitForEnqueue being used, and then uses invokeAny to submit 4 tasks, with the first one failing, the second one blocking, the third one stuck in the queue, and the fourth one waiting for a queue position.  It interrupts the thread that runs invokeAny and expects InterruptedException, however, intermittently the invokeAny operation can return a result.  The problem is in the test case - the third and fourth tasks, if run, are able to complete immediately.  Under the right timing/ordering, when the interrupted invokeAny causes tasks be canceled, the third and or fourth task can start running and complete before invokeAny finishes, allowing invokeAny to see a successful result and return it.  The fix is to make the third and fourth tasks block so that if they ever do start, they will not complete successfully.

--- a/dev/com.ibm.ws.threading_policy_fat/test-applications/basicfat/src/web/PolicyExecutorServlet.java
+++ b/dev/com.ibm.ws.threading_policy_fat/test-applications/basicfat/src/web/PolicyExecutorServlet.java
@@ -3281,7 +3281,7 @@ public class PolicyExecutorServlet extends FATServlet {
     }
 
     // Submit a group of tasks to timed invokeAny, one of which raises an exception, another of which blocks all others from starting,
-    // another of which it stuck in the queue, and another of which is blocked attempting to get a queue position.
+    // another of which is stuck in the queue, and another of which is blocked attempting to get a queue position.
     // Interrupt the invokeAny operation before it completes. Expect InterruptedException.
     @Test
     public void testInvokeAnyTimedInterruptWaitForEnqueue() throws Exception {
@@ -3295,10 +3295,10 @@ public class PolicyExecutorServlet extends FATServlet {
         CountDownLatch unused = new CountDownLatch(0);
         Collection<CountDownTask> tasks = Arrays.asList(new CountDownTask(beginLatch_1_2, null, 0), // Intentionally fail with NullPointerException
                                                         new CountDownTask(beginLatch_1_2, blocker, TIMEOUT_NS * 2), // block
-                                                        new CountDownTask(unused, unused, 0), // stuck in queue
-                                                        new CountDownTask(unused, unused, 0)); // blocked from entering queue
+                                                        new CountDownTask(unused, blocker, TIMEOUT_NS * 2), // stuck in queue until invokeAny is interrupted
+                                                        new CountDownTask(unused, blocker, TIMEOUT_NS * 2)); // blocked from entering queue until invokeAny is interrupted
 
-        // Interrupt the current thread after both tasks start
+        // Interrupt the current thread after the first two tasks start
         Future<?> interrupterFuture = testThreads.submit(new InterrupterTask(Thread.currentThread(), beginLatch_1_2, TIMEOUT_NS * 2, TimeUnit.NANOSECONDS));
 
         long start = System.nanoTime();

